contracts/FractalUsul.sol

```diff
@@ -1,86 +1,469 @@
+    /// @param value ether value to pass with the call
+    /// @param data the data to be executed from the call
+    /// @param operation Call or Delegatecall
+    function executeProposalByIndex(
+        uint256 proposalId,
+        address target,
+        uint256 value,
+        bytes memory data,
+        Enum.Operation operation
+    ) public {
+        require(
+            state(proposalId) == ProposalState.Executable,
+            "Proposal must be in the executable state"
+        );
+        bytes32 txHash = getTxHash(target, value, data, operation);
+        require(
+            proposals[proposalId].txHashes[
+                proposals[proposalId].executionCounter
+            ] == txHash,
+            "Transaction hash does not match the indexed hash"
+        );
+        proposals[proposalId].executionCounter++;
+        require(
+            exec(target, value, data, operation),
+            "Module transaction failed"
+        );
+        emit TransactionExecuted(proposalId, txHash);
+    }
+
+    /// @notice Executes all the transactions within a proposal
+    /// @param proposalId the identifier of the proposal
+    /// @param targets the contracts to be called by the avatar
+    /// @param values ether values to pass with the calls
+    /// @param data the data to be executed from the calls
+    /// @param operations Calls or Delegatecalls
+    function executeProposalBatch(
+        uint256 proposalId,
+        address[] memory targets,
+        uint256[] memory values,
+        bytes[] memory data,
+        Enum.Operation[] memory operations
+    ) external {
+        require(
+            targets.length != 0,
+            "No transactions to execute provided"
+        );
+        require(
+            targets.length == values.length &&
+                targets.length == data.length &&
+                targets.length == operations.length,
+            "Array length mismatch"
+        );
+        require(
+            proposals[proposalId].executionCounter + targets.length <=
+                proposals[proposalId].txHashes.length,
+            "Too many transactions to execute provided"
+        );
+        for (uint256 i = 0; i < targets.length; i++) {
+            executeProposalByIndex(
+                proposalId,
+                targets[i],
+                values[i],
+                data[i],
+                operations[i]
+            );
+        }
+        emit TransactionExecutedBatch(
+            proposals[proposalId].executionCounter,
+            proposals[proposalId].executionCounter + targets.length
+        );
+    }
+
+    /// @notice Enables the specified array of strategy contract addresses
+    /// @param _strategies The array of strategy contract addresses
+    function setupStrategies(address[] memory _strategies) internal {
+        require(
+            strategies[SENTINEL_STRATEGY] == address(0),
+            "setupStrategies has already been called"
```

Comment:
does this mean the available strategies can only be set up once?

Reply:
hmm, I hadn't realized that.  So you can't add a voting strategy to an existing Usul module that has already been set up with one?

I suppose that's not a huge issue currently, but I wonder if there's a way we can allow for strategies to be added after the fact

I could see a DAO starting as a linear token voting DAO, but adding quadratic voting later one.

But maybe in that case they would just have to remove the current module and add a new one with the proper strategy.

Reply:
ah, ok cool, that's good, I think I misread how that works, definitely something we want to be able to do so that's great